CRM/Core/Form.php

```diff
@@ -1514,8 +1514,8 @@ public function addField($name, $props = [], $required = FALSE, $legacyDate = TR
     }
 
     // Core field - get metadata.
-    $fieldSpec = civicrm_api3($props['entity'], 'getfield', $props);
-    $fieldSpec = $fieldSpec['values'];
+    $fieldSpec = civicrm_api3($props['entity'], 'getfield', $props)['values'];
+    $dataType = $fieldSpec['type'];
```

Comment:
3 lines down from here it refers to `$fieldSpec['html']['type']` but this says `$fieldSpec['type']`. One of those must be wrong.

Reply:
Actually $fieldSpec['html']['type'] indicate the HTML widget type and $fieldSpec['type'] denotes the dataType. As in this case the metadata for is_archived field is 

```
Array
(
    [name] => is_archived
    [type] => 16
    [title] => Is Mailing Archived?
    [description] => Is this mailing archived?
    [where] => civicrm_mailing.is_archived
    [table_name] => civicrm_mailing
    [entity] => Mailing
    [bao] => CRM_Mailing_BAO_Mailing
    [localizable] => 0
    [html] => Array
        (
            [type] => Radio
        )

    [is_core_field] => 1
)
```

Reply:
Oh ok.
And is that array key guaranteed to exist? Even for custom fields? Or do we need a guard to prevent undefined index warnings?

Reply:
Yes I think so. Custom fields won't reach there because such fields are rendered before coming to this point (see L1510-1514). Whereas core field's meta-data is always generated from XML which has an essential type attribute. However its the HTML type, which is optional